### DISCUSSION TITLE
Unit test gives conflicting result from interacting with docs

### DISCUSSION THREAD

**Discussion by mwilson8** [fastapi/fastapi#7575] [Questions] [answered]
### Describe the bug

We're developing an application with docker and after building, I can interact with the docs and see that the endpoints behave as expected.. but when running unit tests that emulate the same action, I get an incorrect result

### To Reproduce

1. Application file:

```Python
from fastapi import FastAPI

app = FastAPI()


@router.get("extract/{extraction_id}", response_model=schema.Extraction, status_code=HTTPStatus.OK)
async def get_extraction_by_id(extraction_id: str, {other deps like db, etc}):

    try:
        extraction: Extraction = _query_extraction(db, extraction_id) 
        # for recreating the error, the below may work simpler
        # raise SQLAlchemyError('some txt')
    except SQLAlchemyError as sqla_err:
        raise HTTPException(detail=str(sqla_err), status_code=HTTPStatus.NOT_FOUND)
```

2. Testing file
```Python
def test_bad_id():
    with TestClient(app) as client:

        attempt = client.get('extract/1')

        assert attempt.status_code == HTTPStatus.NOT_FOUND
```

3. Build the project with docker
4. Open the browser and call the endpoint `/extract/1`.

> Sending with an extraction_id of '1' causes a validation error within SQLAlchemy because it's not a valid uuid, so the call to _query_extraction(...) throws an SQLAlchemyError (which is caught and an HTTPError thrown)

- It returns a HTTPError with status code 404 

5. Run the testing file with pytest

- It fails 
```bash
===================================== FAILURES ======================================
_____________________________________test_bad_id_____________________________________ 
    def test_bad_id():
        with TestClient(app) as client:

            attempt = client.get('extract/1', headers=TESTER_10_HEADERS)

>           assert attempt.status_code == HTTPStatus.NOT_FOUND
E           assert <HTTPStatus.OK: 200> == <HTTPStatus.NOT_FOUND: 404>
E            +  where <HTTPStatus.OK: 200> = <Response [HTTPStatus.OK]>.status_code
E            +  and   <HTTPStatus.NOT_FOUND: 404> = HTTPStatus.NOT_FOUND

tests/test_endpoint_db.py:115: AssertionError
```

### Expected behavior

Expected the test to pass with receiving the status code of 404

### Screenshots
<img width="1432" alt="interactive docs" src="https://user-images.githubusercontent.com/14056375/80542868-ce19c700-897b-11ea-913a-c41970391146.png">
<img width="1214" alt="unit test" src="https://user-images.githubusercontent.com/14056375/80542877-d245e480-897b-11ea-8afc-8efa4d334323.png">

### Environment

- osX Catalina locally (Docker I believe alpine python)
- 'fastapi==0.47.1'
- Python 3.6.8

### Additional context

Other tests where we check response codes about validation all pass. This originated after adding an exception handler and noticing the same thing was happening with it.. the interactive docs showed that the exception was caught by the handler and the expected response code was received, but when unit testing it was still giving a 200 OK

**Comment by phy25** [comment]
I would add `print("caught")` under your `except SQLAlchemyError as sqla_err:` line to see if it really went through that block.

If yes, you can simply raise HTTPException in a testing endpoint to see if HTTPException is not working in TestClient. Also how do you import your HTTPException?

**Comment by retnikt** [comment]
Try adding `print(attempt.headers, attempt.body, sep="\n\n")` to before the assertion in your test and post the result please, it might contain some useful information

**Comment by retnikt** [comment]
Also, when you do this are you running the tests locally or in the docker container? What Dockerfile are you using?

**Comment by mwilson8** [comment]
> I would add `print("caught")` under your `except SQLAlchemyError as sqla_err:` line to see if it really went through that block.
> 
> If yes, you can simply raise HTTPException in a testing endpoint to see if HTTPException is not working in TestClient. Also how do you import your HTTPException?

```Python
from fastapi import APIRouter, Depends, HTTPException
```

Yes, it prints "caught", so I changed the endpoint to immediately raise an Exception:
```Python
@router.get("extract/{extraction_id}", response_model=schema.Extraction, status_code=HTTPStatus.OK)
async def get_extraction_by_id(extraction_id: str, {other deps like db, etc}):

    raise HTTPException("some_text", status_code=HTTPStatus.NOT_FOUND)
```



> Try adding `print(attempt.headers, attempt.body, sep="\n\n")` to before the assertion in your test and post the result please, it might contain some useful information

`attempt.body` was an unrecognized attribute, but `attempt.content` showed some bogus json of the model we use for the return type.. 
```bash
{'content-length': '1004', 'content-type': 'application/json'}

b'{"created_by":"^","updated_by":"b(3Gr\\u000b/#","created_at":"8072-04-08T23:35:51.118267","updated_at":"7407-08-14T08:44:23.557034","deleted_at":"8242-11-16T09:38:39.498934","deleted_by":"w","extraction_id":"6513270e-269e-4d37-b2a7-4de452e6b438","uri":"vP2!pFzEA |\\n %!7\\r0*W\\faq\\"\\n\\rn\\r{","protocol":"","location":"S[]~\\"(Bm=C","meta":{"content_type":"N+=","parsers":["N+=","N+="],"parse_time_millis":"N+=","resource_name":"N+=","raw_meta":"N+="},"content":null,"status":"processing","checksum":""}'
```


> Also, when you do this are you running the tests locally or in the docker container? What Dockerfile are you using?

within docker
```bash
$ docker-compose up --build -d
$ docker-compose exec akit /bin/sh
/usr/src/app # python -m pytest
```

```Dockerfile
FROM <CDN image> essentially FROM python:3.6-alpine3.6

COPY . /usr/src/app
WORKDIR /usr/src/app

ARG CI

ENV CI ${CI}

# Tika specific settings
ENV TIKA_CLIENT_ONLY True
ENV TIKA_VERSION 1.19
ENV TIKA_SERVER_JAR /usr/src/app/pkg/jars/tika-server.jar
ENV TIKA_LOG_PATH /usr/src/app/logs
ENV TIKA_PATH /usr/src/app/pkg/jars

# pip command
ARG PIP_CMD='pip install --no-cache-dir'

# Alpine package command
ARG APK_ADD='apk add --no-cache'

# dependencies that will be added to the image before pip is run and remain in the image
ARG PKGS='git libpq openssh-client openjdk11 curl'

# development packages that will be installed before pip is run and purged after
ARG DEV_PKGS='linux-headers postgresql-dev python3-dev gcc g++ musl-dev libpng-dev libressl-dev libffi-dev build-base'

RUN cp /usr/src/app/docker/akit/ssh/* /root/.ssh && chmod -R 600 /root/.ssh && \
    if test -n "$CI"; then sed -i '/#    Hostname.*/s/^#//' /root/.ssh/config; fi && \
    $PIP_CMD pip --upgrade && \
    pip uninstall -y pymera && \
    $APK_ADD openjdk11 --repository=http://dl-cdn.alpinelinux.org/alpine/edge/community && \
    if ! $PIP_CMD -r requirements.txt; then $APK_ADD $PKGS && $APK_ADD --virtual .build-deps $DEV_PKGS && $PIP_CMD -r requirements.txt && apk del .build-deps; fi && \
    $PIP_CMD gunicorn && $PIP_CMD uvicorn && \
    rm -r /root/.ssh/id_rsa* && \
    mkdir -p /tmp

ENTRYPOINT ["sh", "/usr/src/app/docker/akit/docker-entrypoint.sh"]

CMD ["uvicorn", \
    "akit:app", \
    "--host", "0.0.0.0", \
    "--port", "8000"]
```

`docker-entrypoint.sh` does some waiting/healthchecks for services and alembic migrations then at the end 
```bash
# execute CMD
if [ -z "$DIR_TO_RELOAD" ]; then
  exec "$@"
else
  exec "$@" "--reload" "--reload-dir" "$DIR_TO_RELOAD"
fi
```

Unfortunately, I'm still getting the same conflicting results

**Comment by retnikt** [comment]
Have you tried changing `extract/{extraction_id}` to `/extract/{extraction_id}` and likewise in the TestClient? I'm not entirely sure if/why that would help, but it seems unusual

**Comment by mwilson8** [comment, marked as answer]
I believe I've found the problem, unrelated to FastAPI

We use Mock to build a fake database to query, and it seems that any tests run after the Mock database has been created will forgo the endpoint logic and just jump straight to querying the database (this was implemented by another team so it took me a while to find it)

Moving the tests for invalid id & valid but with empty db to above Mock database creation produced expected results

Closing the issue for now, thank you for your time!

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue :+1:
